### PR TITLE
Medium: IPsrcaddr: Add "update_local_route" attribute

### PR DESCRIPTION
**Pull request by nrwahl2**
Currently, the IPsrcaddr resource agent updates the default route to
use a specified IP address (`OCF_RESKEY_ipaddress`) as the preferred
source IP. It does not make any changes to the local route table.

This patch adds an "update_local_route" boolean attribute
(default=false), which updates the local route for destination
`$OCF_RESKEY_ipaddress` to use `src $OCF_RESKEY_ipaddress`.

#### Rationale
`IPsrcaddr` is often paired with `IPaddr2` to provide a virtual IP
address and then use that virtual IP as the preferred source IP for
outgoing connections on the default route. Some applications make local
connections to the virtual IP address. A subset of those require
that the source IP for the connection be the virtual IP.

In other words, the local connection in these cases needs to be:
```
    virtual_ip -> virtual_ip
```

However, this is what is occurring:
```
    primary_ip -> virtual_ip
```

The local routing table determines the source IP for these connections.
For routes in the local routing table, the default src IP address has
always been the primary IP address of the interface. After adding an
`IPaddr2` resource, there will be a local route like this:

```
# ip route show table local to exact <virtual_ip>
local <virtual_ip> dev <iface> proto kernel scope host src <primary_ip>
```

The IPsrcaddr resource agent needs to be able to update the local route
to `OCF_RESKEY_ipaddress` in addition to updating the default route.

### CODE REVIEW

**Comment by nrwahl2** [comment]
I have added only that which is necessary for the local route update, plus replacing `ip` with `$IP2UTIL` in the `NETWORK` variable assignment. There are some other formatting, consistency, and safe parsing issues in the RA that should also be addressed, but I will keep it simple for now. I've stylistically aligned the new additions with the existing content as best I can.

**Comment by oalbrigt** [comment]
add to whitelist

**Comment by nrwahl2** [comment]
Related to RHBZ#1582685

File: heartbeat/IPsrcaddr [hunk 1]
```diff
@@ -177,7 +197,14 @@ srca_start() {
 
 		$CMDCHANGE $ROUTE_WO_SRC src $1 || \
```
Comment (oalbrigt) [inline heartbeat/IPsrcaddr]: Wouldnt it be either this command orthe one in the if-statement?

Comment (nrwahl2) [reply heartbeat/IPsrcaddr]: The `if` statement only includes an `ocf_log_info`. Maybe you meant either this command or the `ip route replace` before it.

```
175                 ip route replace $NETWORK dev $INTERFACE src $1 || \
176                         errorexit "command 'ip route replace $NETWORK dev $INTERFACE src $1' failed"
177 
178                 $CMDCHANGE $ROUTE_WO_SRC src $1 || \
179                         errorexit "command '$CMDCHANGE $ROUTE_WO_SRC src $1' failed"
```

I'm not sure what the effective difference is between lines 175 and 178 in the original version. I've left the existing stuff alone and just added what's needed for `update_local_route=true` to work.

File: heartbeat/IPsrcaddr [hunk 2]
```diff
@@ -177,7 +197,14 @@ srca_start() {
...
 			errorexit "command '$CMDCHANGE $ROUTE_WO_SRC src $1' failed"
+
 		rc=$?
```
Comment (oalbrigt) [inline heartbeat/IPsrcaddr]: You might want to return early if it fails here.

File: heartbeat/IPsrcaddr (hunk 1 above)
Comment (oalbrigt) [reply heartbeat/IPsrcaddr]: You're welcome to update to $IP2UTIL.

File: heartbeat/IPsrcaddr (hunk 2 above)
Comment (oalbrigt) [inline heartbeat/IPsrcaddr]: You should return early if it fails here.

File: heartbeat/IPsrcaddr (hunk 1 above)
Comment (oalbrigt) [reply heartbeat/IPsrcaddr]: You're welcome to update it to $IP2UTIL.

File: heartbeat/IPsrcaddr (hunk 2 above)
Comment (nrwahl2) [reply heartbeat/IPsrcaddr]: The `errorexit` does that. Custom function that the creator of this RA wrote.

```
errorexit() {
        ocf_exit_reason "$*"
        exit $OCF_ERR_GENERIC
}
```

File: heartbeat/IPsrcaddr (hunk 1 above)
Comment (nrwahl2) [reply heartbeat/IPsrcaddr]: Will do!

**Comment by nrwahl2** [comment]
This is going to need some more attention, to add a flag for `update_default_route` so that users can configure multiple local routes.

File: heartbeat/IPsrcaddr (hunk 2 above)
Comment (oalbrigt) [reply heartbeat/IPsrcaddr]: Ah. I didnt think of that :)

**Comment by nrwahl2** [comment]
Closed in favor of https://github.com/ClusterLabs/resource-agents/pull/1212.
